processing/src/main/java/org/apache/druid/segment/join/table/BroadcastSegmentIndexedTable.java

```diff
@@ -0,0 +1,255 @@
+        false,
+        null
+    );
+
+    final Sequence<Integer> sequence = Sequences.map(
+        cursors,
+        cursor -> {
+          if (cursor == null) {
+            return 0;
+          }
+          int rowNumber = 0;
+          ColumnSelectorFactory columnSelectorFactory = cursor.getColumnSelectorFactory();
+
+          // this should really be optimized to use dimension selectors where possible to populate indexes from bitmap
+          // indexes, but, an optimization for another day
+          final List<BaseObjectColumnValueSelector> selectors = keyColumnNames
+              .stream()
+              .map(columnName -> {
+                // multi-value dimensions are not currently supported
+                if (adapter.getColumnCapabilities(columnName).hasMultipleValues().isMaybeTrue()) {
+                  return NilColumnValueSelector.instance();
+                }
+                return columnSelectorFactory.makeColumnValueSelector(columnName);
+              })
+              .collect(Collectors.toList());
+
+          while (!cursor.isDone()) {
+            for (int keyColumnSelectorIndex = 0; keyColumnSelectorIndex < selectors.size(); keyColumnSelectorIndex++) {
+              final String keyColumnName = keyColumnNames.get(keyColumnSelectorIndex);
+              final int columnPosition = rowSignature.indexOf(keyColumnName);
+              final Map<Object, IntList> keyColumnValueIndex = keyColumnsIndex.get(columnPosition);
+              final Object key = selectors.get(keyColumnSelectorIndex).getObject();
+              if (key != null) {
+                final IntList array = keyColumnValueIndex.computeIfAbsent(key, k -> new IntArrayList());
+                array.add(rowNumber);
+              }
+            }
+
+            if (rowNumber % 100_000 == 0) {
+              if (rowNumber == 0) {
+                LOG.debug("Indexed first row for table %s", theSegment.getId());
+              } else {
+                LOG.debug("Indexed row %s for table %s", rowNumber, theSegment.getId());
+              }
+            }
+            rowNumber++;
+            cursor.advance();
+          }
+          return rowNumber;
+        }
+    );
+
+    Integer totalRows = sequence.accumulate(0, (accumulated, in) -> accumulated += in);
+    LOG.info("Created BroadcastSegmentIndexedTable with %s rows.", totalRows);
+  }
+
+  @Override
+  public String version()
+  {
+    return version;
+  }
+
+  @Override
+  public Set<String> keyColumns()
+  {
+    return keyColumns;
+  }
+
+  @Override
+  public RowSignature rowSignature()
+  {
+    return rowSignature;
+  }
+
+  @Override
+  public int numRows()
+  {
+    return adapter.getNumRows();
+  }
+
+  @Override
+  public Index columnIndex(int column)
+  {
+    return RowBasedIndexedTable.getKeyColumnIndex(column, keyColumnsIndex, rowSignature);
+  }
+
+  @Override
+  public Reader columnReader(int column)
+  {
+    if (column < 0 || rowSignature.getColumnName(0) == null) {
```

Comment:
```suggestion
    if (column < 0 || rowSignature.getColumnName(column) == null) {
```

Reply:
oops, good catch, though it accidentally will throw the same index out of bounds exception a few lines later

```java
    final BaseColumn baseColumn = queryableIndex.getColumnHolder(rowSignature.getColumnName(column)).getColumn();
```

However, this check isn't really a good check since it is inconsistent which exception will be thrown, so I've added a `contains` method to `RowSignature` that works with column position to ensure that the index is in range so the exception can be consistently be the IAE being thrown here.